### PR TITLE
Status Effect & Potion API

### PR DESCRIPTION
**Pull request by Boundarybreaker**
Not yet finished. Currently needs:
- [x] `StatusEffects` class for all vanilla status effects
- [x] `Potions` class for all vanilla potions
- [ ] Discuss: zero-index or one-index amplifiers?
- [x] Discuss: how to do strong/long `Potion` variants (hardcode register or add system for modifying based on item NBT)
- [ ] Potion util class
- [x] Potion item class?

### CODE REVIEW

File: src/main/java/org/sandboxpowered/sandbox/api/status/StatusEffect.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import javax.annotation.Nullable;
+
+public interface StatusEffect {
+
+    StatusEffectType getType();
+
+    void applyUpdateEffect(LivingEntity entity, int amplifier);
+
+    //TODO: double-check damage source to make sure the thrower and the potion are named right
+    void applyInstantEffect(@Nullable Entity thrower, @Nullable Entity potion, LivingEntity target, int amplifier, double distance);
```
Comment (TheCodedOne) [inline src/main/java/org/sandboxpowered/sandbox/api/status/StatusEffect.java]: I'd prefer Mono's over Nullable here

**Comment by Boundarybreaker** [comment]
Hm. I've been looking over potion stuff again, and I don't think it's worth it to have different potion lengths/strengths defined within a single `Potion` object. It'd just make it far too much of a nightmare for dealing with how the brewing stand works. Maybe we could also have a `PotionCollection` class that can do the registry stuff for a potion and all of its variants.

**Comment by Boundarybreaker** [comment]
Alright, the `PotionSet` class is a helper for potions with long/strong variants, which will auto-register any long/strong/etc. variants you want (with the ability to define your own variant set). Vanilla potions are provided as themselves if they have no variants, or `PotionSet`s if they do have variants. Splash and Lingering potions are handled in a completely different manner, so we luckily don't have to worry about them much.

Next, I'm gonna make a PotionUtil class and see what would be needed to hook into potion brewing.

**Comment by Boundarybreaker** [comment]
This needs to be updated for the new Functions system and it probably needs a potion util class, but otherwise it's good to go I think?

**Comment by Boundarybreaker** [comment]
Gonna move this to be built on top of 0.4.

**Pull request closed without merging**